encoding/thrift/inbound_nowire.go

```diff
@@ -0,0 +1,207 @@
+)
+
+var _emptyResponse = NoWireResponse{}
+
+// thriftNoWireHandler is similar to thriftUnaryHandler and thriftOnewayHandler
+// except that thriftNoWireHandler implements both transport.UnaryHandler and
+// transport.OnewayHandler through a single type, utilizing the "nowire"
+// (streaming in the ThriftRW context, bypassing the intermediate "Wire"
+// representation) implementation.
+//
+// The UnaryHandler and OnewayHandler implementations then call into type that
+// implements NoWireHandler, which understands how to unpack and invoke the
+// request, given the ThriftRW primitives for reading the raw representation.
+type thriftNoWireHandler struct {
+	NoWireHandler NoWireHandler
+	Protocol      stream.Protocol
+	Enveloping    bool
+}
+
+var (
+	_ transport.OnewayHandler = (*thriftNoWireHandler)(nil)
+	_ transport.UnaryHandler  = (*thriftNoWireHandler)(nil)
+)
+
+// NoWireCall contains all of the required objects needed for an underlying
+// Handle needs to unpack any given request.
+type NoWireCall struct {
+	Reader        io.Reader
+	RequestReader stream.RequestReader
+	EnvelopeType  wire.EnvelopeType
+}
+
+// NoWireHandler is implemented by the generated code for each method that the
+// server needs to implement.  It is responsible for unpacking the request,
+// executing it, and returning a NoWireResponse that contains information about
+// how to construct a response as well as any relevant metadata while executing
+// the request.
+type NoWireHandler interface {
+	Handle(context.Context, *NoWireCall) (NoWireResponse, error)
+}
+
+func (t thriftNoWireHandler) Handle(ctx context.Context, treq *transport.Request, rw transport.ResponseWriter) error {
+	ctx, call := encodingapi.NewInboundCall(ctx)
+	defer closeReader(treq.Body)
+
+	res, err := decodeNoWireRequest(ctx, call, treq, rw, t.NoWireHandler, wire.Call, t.Protocol, t.Enveloping)
+	if err != nil {
+		return err
+	}
+
+	if resType := res.Body.EnvelopeType(); resType != wire.Reply {
+		return errors.ResponseBodyEncodeError(
+			treq, errUnexpectedEnvelopeType(resType))
+	}
+
+	if res.IsApplicationError {
+		rw.SetApplicationError()
+
+		if applicationErrorMetaSetter, ok := rw.(transport.ApplicationErrorMetaSetter); ok {
+			applicationErrorMetaSetter.SetApplicationErrorMeta(&transport.ApplicationErrorMeta{
+				Details: res.ApplicationErrorDetails,
+				Name:    res.ApplicationErrorName,
+				Code:    res.ApplicationErrorCode,
+			})
+		}
+	}
+
+	if err := call.WriteToResponse(rw); err != nil {
+		// not reachable
+		return err
+	}
+
+	if err := res.ResponseWriter.WriteResponse(wire.Reply, rw, res.Body); err != nil {
+		return errors.ResponseBodyEncodeError(treq, err)
+	}
+
+	return nil
+}
+
+func (t thriftNoWireHandler) HandleOneway(ctx context.Context, treq *transport.Request) error {
+	ctx, call := encodingapi.NewInboundCall(ctx)
+	defer closeReader(treq.Body)
+
+	_, err := decodeNoWireRequest(ctx, call, treq, nil, t.NoWireHandler, wire.OneWay, t.Protocol, t.Enveloping)
```

Comment:
That's a lot of shared parameters that are just fields of t. I would make this a method of t.

Fixing.